### PR TITLE
Add support for memcpy_s

### PR DESCRIPTION
**Pull request by hernanponcedeleon**
Somehow, I ended up closing #685 and I don't find a way to open it again.

This PR should give a proper implementation of `memcpy_s`, including the runtime checks to avoid UB.

### CODE REVIEW

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java [hunk 1]
```diff
@@ -1328,6 +1329,86 @@ private List<Event> inlineMemCpy(FunctionCall call) {
...
+            // FIXME: We have no other choice but to load ptr-sized chunks for now
+            final Register reg = caller.getOrNewRegister("__memcpy_s_" + i, types.getArchType());
+            // We should zero out the entire destination range [dest, dest+destsz)
+            // Here we handle [dest, dest+count), the remaining are handled by the loop below
+            final Expression value = expressions.makeITE(c4, expressions.makeZero(types.getArchType()), reg);
+
+            replacement.addAll(List.of(
+                    EventFactory.newLoad(reg, srcAddr),
+                    EventFactory.newStore(destAddr, value)
+            ));
```
Comment (ThomasHaas) [inline dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java:1384]: There are three cases for `memcpy_s`:
- The inputs are valid and you do a standard memcpy (+ generate a success error code).
- The inputs are invalid and in particular `dest == NULL`, then the method only produces an error code without any memory accesses.
- The inputs are invalid but at least `dest != NULL` (*), then the range `[dest, dest + destsz]` gets zeroed which only requires memory stores (no loads).

I don't think your implementation is correct. You have to use proper branching to make the case distinctions rather than trying to use ITE.

Btw. how is this function even used in practice? Does the compiler magically transform `memcpy` to `memcpy_s` or does the user specficially call the latter? If so, to what extent can the user recover from an error or does he just terminate the program but with an error code that is better than segfault?
Basically, I'm wondering if failure of `memcpy_s` is a bug in pretty much all cases. If so, we could also generate an `assert (boundCheck) ### memcpy_s error` and don't even bother too much about the failure details (we could also provide an option to toggle this behaviour).

(*) Technically, `destsz` could also be invalid but only if it is larger than `RSIZE_MAX` which is impossible in the implementation anyway.

**Comment by ThomasHaas** [comment]
I think we should have a unit test related to `memcpy_s`. This one seems to be not so straightforward.

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java [hunk 2]
```diff
@@ -1328,6 +1329,86 @@ private List<Event> inlineMemCpy(FunctionCall call) {
...
+        final Expression srcIsNull = expressions.makeEQ(src, nullExpr);
+        // We assume RSIZE_MAX = 2^64-1
+        final Expression rsize_max = expressions.makeValue(BigInteger.ONE.shiftLeft(64).subtract(BigInteger.ONE), types.getArchType());
+        final Expression gtMax = expressions.makeOr(
+                expressions.makeGT(countExpr, rsize_max, false),
+                expressions.makeGT(destszExpr, rsize_max, false));
+        final Expression countGtdestszExpr = expressions.makeGT(countExpr, destszExpr, false);
+        final Expression overlap = expressions.makeOr(
+                expressions.makeGTE(expressions.makeAdd(src, countExpr), dest, false),
+                expressions.makeGTE(expressions.makeAdd(dest, countExpr), src, false));
```
[the remark this answers is not in the document]
Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java:1362]: Good catch. I also think it would be correct if we use `makeAnd`. I will go with this.

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java (hunk 1 above)
Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java:1384]: Ok, I missed the "(if both dest and destsz are valid)". After re-reading, this is my understanding
- `dest == NULL` or `invalid destsz` ----> return error > 0
- `dest != NULL && valid destsz && (src == NULL || count > destsz || overlap(src, dest))` ----> return error > 0 and zero out `[dest, dest+destsz)`

> I don't think your implementation is correct. You have to use proper branching to make the case distinctions rather than trying to use ITE.

Why do you think we need branches? I don't see this.

> Btw. how is this function even used in practice? Does the compiler magically transform `memcpy` to `memcpy_s` or does the user specficially call the latter? 

The user explicitly uses the later. This is used in libvsync and the whole reason of this PR. It is then the responsibility of the developer to check if the returned value != 0.

> If so, to what extent can the user recover from an error or does he just terminate the program but with an error code that is better than segfault? Basically, I'm wondering if failure of `memcpy_s` is a bug in pretty much all cases. If so, we could also generate an `assert (boundCheck) ### memcpy_s error` and don't even bother too much about the failure details (we could also provide an option to toggle this behaviour).

This is up to the user. If we want to be precise about its semantics, I don't think we can make this.
If we don't want to be precise, I would simply treat `memcpy_s` as `memcpy` and get minimal changes to the current code.

> 
> (*) Technically, `destsz` could also be invalid but only if it is larger than `RSIZE_MAX` which is impossible in the implementation anyway.

Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java:1384]: > Ok, I missed the "(if both dest and destsz are valid)". After re-reading, this is my understanding
> 
>     * `dest == NULL` or `invalid destsz` ----> return error > 0
> 
>     * `dest != NULL && valid destsz && (src == NULL || count > destsz || overlap(src, dest))` ----> return error > 0 and zero out `[dest, dest+destsz)`

Yes, that seems right. Importantly, in neither of those two cases you load from `src`!

> 
> 
> > I don't think your implementation is correct. You have to use proper branching to make the case distinctions rather than trying to use ITE.
> 
> Why do you think we need branches? I don't see this.

You just listed the different cases. One case performs loads and stores, another only stores, and the third one has no memory accesses. There is no way to get this behaviour without three different control-flow branches.
Your solution tried to merge cases (1) and (2) by always performing the loads and just ignoring their results to mimic case (2). These redundant loads are already questionable and you should avoid them, but you definitely cannot have redundant stores so case (3) must be different than the other two.

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java [hunk 3]
```diff
@@ -1328,6 +1331,108 @@ private List<Event> inlineMemCpy(FunctionCall call) {
...
+        final Expression srcIsNull = expressions.makeEQ(src, nullExpr);
+        // We assume RSIZE_MAX = 2^64-1
+        final Expression rsize_max = expressions.makeValue(BigInteger.ONE.shiftLeft(64).subtract(BigInteger.ONE), types.getArchType());
+        final Expression invalidDestsz = expressions.makeGT(
+                expressions.makeCast(destszExpr, types.getArchType()), rsize_max, false);
+        final Expression countGtMax = expressions.makeGT(
+                expressions.makeCast(countExpr, types.getArchType()), rsize_max, false);
+        final Expression countGtdestszExpr = expressions.makeGT(countExpr, destszExpr, false);
+        final Expression invalidCount = expressions.makeOr(countGtMax, countGtdestszExpr);
+        // src and dest are pointers (int64), count is usually interpreted as int32, thus the cast
```
Comment (hernanponcedeleon) [inline dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java:1367]: This and the other casts in the pass really bother me. Without them, I was getting different BVs sizes, but I think the compiler and the type system should take care of giving me the correct types. So there is something that still seems fishy here. I suspect my system might not be properly recognizing `memcpy_s`

Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java:1367]: You can see this problem in your `.ll` file:
```
declare dso_local i32 @memcpy_s(...) #3 // varargs declaration because the function is undeclared in the C-code
//... The call is thus a varags call with ugly function type casts:
  %73 = call i32 (i32*, i64, i32*, i64, ...) bitcast (i32 (...)* @memcpy_s to i32 (i32*, i64, i32*, i64, ...)*)(i32* %71, i64 12, i32* %72, i64 12), !dbg !86
```
This function is declared in `string.h` according to cppreference and you are not including that one.

Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java:1367]: Even if I add `string.h`, function `memcpy_s` ends up with an empty body in the llvm file. There is this

> As with all bounds-checked functions, memcpy_s only guaranteed to be available if __STDC_LIB_EXT1__ is defined by the implementation and if the user defines __STDC_WANT_LIB_EXT1__ to the integer constant 1 before including [<string.h>](https://en.cppreference.com/w/c/string/byte).

and by trying the code from [here](https://stackoverflow.com/questions/47867130/stdc-lib-ext1-availability-in-gcc-and-clang), I get `#error "__STDC_LIB_EXT1__ not defined"`, so we might need a solution that works even if the implementation of `memcpy_s` is not available.

However, if I remove the casts from the code, I am now not able to reproduce the parsing problems I was getting before. What I don't understand is that even if the llvm code show different types in the call arguements
```
  %8 = call i32 (i8*, i64, i32*, i64, ...) bitcast (i32 (...)* @memcpy_s to i32 (i8*, i64, i32*, i64, ...)*)(i8* null, i64 12, i32* %7, i64 12), !dbg !27
  %15 = call i32 (i32*, i64, i8*, i64, ...) bitcast (i32 (...)* @memcpy_s to i32 (i32*, i64, i8*, i64, ...)*)(i32* %14, i64 12, i8* null, i64 12), !dbg !34
  %44 = call i32 (i32*, i64, i32*, i64, ...) bitcast (i32 (...)* @memcpy_s to i32 (i32*, i64, i32*, i64, ...)*)(i32* %42, i64 8, i32* %43, i64 16), !dbg !60
  %73 = call i32 (i32*, i64, i32*, i64, ...) bitcast (i32 (...)* @memcpy_s to i32 (i32*, i64, i32*, i64, ...)*)(i32* %71, i64 12, i32* %72, i64 12), !dbg !86
  %102 = call i32 (i32*, i64, i32*, i64, ...) bitcast (i32 (...)* @memcpy_s to i32 (i32*, i64, i32*, i64, ...)*)(i32* %100, i64 12, i32* %101, i64 12), !dbg !112
```
we parse all of them as bv64
```
11:      bv32 r8 <- call memcpy_s(bv64(0), bv64(12), bv64 r7, bv64(12))
30:      bv32 r15 <- call memcpy_s(bv64 r14, bv64(12), bv64(0), bv64(12))
107:     bv32 r44 <- call memcpy_s(bv64 r42, bv64(8), bv64 r43, bv64(16))
184:     bv32 r73 <- call memcpy_s(bv64 r71, bv64(12), bv64 r72, bv64(12))
261:     bv32 r102 <- call memcpy_s(bv64 r100, bv64(12), bv64 r101, bv64(12))
```

Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java]: > The count/size parameters are more dangerous. Since your code uses `sizeof(...)` everywhere, you will get a consistent type `size_t` which equals 64 bit on most systems. If you were to put constants that are interpreted as 32 bit ints, then you would see the problem again. 

The type of those is `rsize_t` which AFAIK is defined as `typedef size_t rsize_t;`. Thus, the compiled should take care of the casting even if constant are used. However, for 32 bits archs, we probably get into problems.

> Always upcasting those parameters to 64 bit is probably fine though.

Since we overall assume 64 bits arch, I guess we should use the casts. I will put those two back.

> 
> EDIT: Btw. even in your newest .ll file, the function is not properly recognized.

If you mean that we have `declare dso_local i32 @memcpy_s(...) #3` with no body, I am aware of this. This is what I meant above with `"__STDC_LIB_EXT1__ not defined"` and apparently is not mandatory to have them implemented

> These functions are defined in ISO 9899:2011 Annex K and they are optional to implement. Before using them you must check if __STDC_LIB_EXT1__ is defined.

Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java]: I'm not talking specifically about the body, but the whole signature. Since you had a commit to include `string.h` and to regenerate the llvm file, I thought you found a way to get this function recognized.
It might make sense you just declare the function in the benchmark to get the right signature (not strictly necessary though).

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java [hunk 4]
```diff
@@ -1328,6 +1331,106 @@ private List<Event> inlineMemCpy(FunctionCall call) {
...
+        // This parameter has type rsize_t/size_t which we model as types.getArchType(), thus the cast
+        final Expression destszExpr = expressions.makeCast(call.getArguments().get(1), types.getArchType());
+        final Expression src = call.getArguments().get(2);
+        // This parameter has type rsize_t/size_t which we model as types.getArchType(), thus the cast
+        final Expression countExpr = expressions.makeCast(call.getArguments().get(3), types.getArchType());
+
+        if (!(countExpr instanceof IntLiteral countValue)) {
+            final String error = "Cannot handle memcpy_s with dynamic count argument: " + call;
+            throw new UnsupportedOperationException(error);
+        }
```
Comment (ThomasHaas) [inline dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java:1349]: This is not going to work. You generate the cast expression and then check for it to be a literal, which should always fail. The reason it works right now is that the cast is recognized as NOOP and optimized away immediately.

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java [hunk 5]
```diff
@@ -1328,6 +1331,106 @@ private List<Event> inlineMemCpy(FunctionCall call) {
...
+        Label check1 = EventFactory.newLabel("__memcpy_s_check_1");
+        Label check2 = EventFactory.newLabel("__memcpy_s_check_2");
+        Label success = EventFactory.newLabel("__memcpy_s_success");
+        Label end = EventFactory.newLabel("__memcpy_s_end");
+
+        // dest == NULL or destsz > RSIZE_MAX ----> return error > 0
+        final Expression c1 = expressions.makeOr(destIsNull, invalidDestsz);
+        CondJump skipE1 = EventFactory.newJump(expressions.makeNot(c1), check2);
+        CondJump skipRest1 = EventFactory.newGoto(end);
+        Local retError1 = EventFactory.newLocal(resultRegister, expressions.makeOne((IntegerType)resultRegister.getType()));
```
Comment (ThomasHaas) [inline dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java:1382]: I think you should refactor this and avoid the magic constants:
```
Expression errorCodeFail = expressions.makeOne((IntegerType)resultRegister.getType());
Expression errorCodeSuccess = expressions.makeZero(...);
```

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java [hunk 6]
```diff
@@ -1328,6 +1331,106 @@ private List<Event> inlineMemCpy(FunctionCall call) {
...
+            check1,
+            skipE1,
+            retError1,
+            skipRest1));
+
+        // dest != NULL && destsz <= RSIZE_MAX && (src == NULL || count > destsz || overlap(src, dest)) 
+        // ----> return error > 0 and zero out [dest, dest+destsz)
+        // The first two are guaranteed by not matching c1
+        final Expression c2 = expressions.makeOr(srcIsNull, invalidCount);
+        final Expression c3 = expressions.makeOr(c2, overlap);
```
Comment (ThomasHaas) [inline dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java:1393]: I would combine these into one line. The intermediary `c2` is confusing. I intuitively thought that it designates a case like `c1` and `c3` but it was just an intermediary step to construct `c3`.
Also, you might want to rename those, cause `cx` could mean `condition x` or `case x` (if one thinks there are multiple cases to check for).
Maybe just have `c1 = condCase1` and `c3 = caseCond2 (getting rid of the intermediate c2)` and then just add a prefix to the comments: `Case 1: dest == null || ...`, `Case 2: dest != null && ...`, `Case 3: All inputs are valid`
I think that is easier to follow with one condition expression per case, and the fact that there are three cases to consider.

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java [hunk 7]
```diff
@@ -1328,6 +1331,106 @@ private List<Event> inlineMemCpy(FunctionCall call) {
...
+            final Register reg = caller.getOrNewRegister("__memcpy_" + i, types.getArchType());
+
+            replacement.addAll(List.of(
+                    EventFactory.newLoad(reg, srcAddr),
+                    EventFactory.newStore(destAddr, reg)
+            ));
+        }
+        replacement.addAll(List.of(
+            retSuccess,
+            end));
```
Comment (ThomasHaas) [inline dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java:1429]: Maybe it is just me, but I like to put the last brackets on a separate line:
```Java
       replacement.addAll(List.of(
            retSuccess,
            end
       ));
```
It's more symmetrical and matches better with how the standard curly brackets are placed.

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java [hunk 8]
```diff
@@ -1328,6 +1331,116 @@ private List<Event> inlineMemCpy(FunctionCall call) {
...
+            check2,
+            skipE2
+        ));
+        for (int i = 0; i < destsz; i++) {
+            final Expression offset = expressions.makeValue(i, types.getArchType());
+            final Expression destAddr = expressions.makeAdd(dest, offset);
+            final Expression value = expressions.makeZero(types.getArchType());
+            replacement.addAll(List.of(
+                EventFactory.newStore(destAddr, value)
+            ));
```
Comment (ThomasHaas) [inline dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java:1414]: Just call `replacement.add`

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java [hunk 9]
```diff
@@ -1328,6 +1331,116 @@ private List<Event> inlineMemCpy(FunctionCall call) {
...
+        CondJump skipRest2 = EventFactory.newGoto(end);
+        Local retError2 = EventFactory.newLocal(resultRegister, errorCodeFail);
+        replacement.addAll(List.of(
+            check2,
+            skipE2
+        ));
+        for (int i = 0; i < destsz; i++) {
+            final Expression offset = expressions.makeValue(i, types.getArchType());
+            final Expression destAddr = expressions.makeAdd(dest, offset);
+            final Expression value = expressions.makeZero(types.getArchType());
```
Comment (ThomasHaas) [inline dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java:1411]: No need to be so indirect and call it `value`, just call it `zero`. It's not a big deal but I think everytime I saw the generated store I intuitively looked for `value` to see what gets stored and everytime it was just zero :).

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java [hunk 10]
```diff
@@ -1328,6 +1331,116 @@ private List<Event> inlineMemCpy(FunctionCall call) {
...
+        replacement.addAll(List.of(
+            check1,
+            skipE1,
+            retError1,
+            skipRest1
+        ));
+
+        // Condition 2: dest != NULL && destsz <= RSIZE_MAX && (src == NULL || count > destsz || overlap(src, dest)) 
+        // ----> return error > 0 and zero out [dest, dest+destsz)
+        // The first two are guaranteed by not matching c1
```
Comment (ThomasHaas) [inline dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java:1399]: "c1" -> "cond1"

**Review by ThomasHaas — APPROVED** [review]
LGTM
